packages/sync/src/modules/Full_Sync.php

```diff
@@ -700,4 +764,57 @@ private function read_option( $name, $default = null ) {
 
 		return $value;
 	}
+
+	/**
+	 * Prefix of the blog lock transient.
+	 *
+	 * @access public
+	 *
+	 * @var string
+	 */
+	const ENQUEUE_LOCK_TRANSIENT_PREFIX = 'jp_sync_enqueue_lock_';
+	/**
+	 * Lifetime of the blog lock transient.
+	 *
+	 * @access public
+	 *
+	 * @var int
+	 */
+	const ENQUEUE_LOCK_TRANSIENT_EXPIRY = 15; // Seconds.
+
+	/**
+	 * Attempt to lock enqueueing when the server receives concurrent requests from the same blog.
+	 *
+	 * @access public
+	 *
+	 * @param int $expiry  enqueue transient lifetime.
+	 * @return boolean True if succeeded, false otherwise.
+	 */
+	public function attempt_enqueue_lock( $expiry = self::ENQUEUE_LOCK_TRANSIENT_EXPIRY ) {
+		$transient_name = $this->get_concurrent_enqueue_transient_name();
+		$locked_time    = get_site_transient( $transient_name );
+		if ( $locked_time ) {
+			return false;
+		}
+		set_site_transient( $transient_name, microtime( true ), $expiry );
+		return true;
+	}
+	/**
+	 * Retrieve the enqueue lock transient name for the current blog.
+	 *
+	 * @access public
+	 *
+	 * @return string Name of the blog lock transient.
+	 */
+	private function get_concurrent_enqueue_transient_name() {
+		return self::ENQUEUE_LOCK_TRANSIENT_PREFIX . get_current_blog_id();
```

Comment:
if we store the transient in a site option that contains the blog ID, should we just save it as a normal option and not contain the blog ID in the transient name?

Reply:
I copied this from the `Server` "lock", I don't remember the reasons why we opted to do it this way, going to investigate.

Reply:
+1 for not using site transients, in favor of regular transients. Having a global-level lock on large / busy multisites sounds like an open door to bugs and contention. If each subsite syncs independently (which they do), I don't see any reason it needs to be a site transient?

Reply:
Also, is a potentially in-memory-only lock durable enough for this purpose? Should this be an option, so that it's backed by the DB (yet still cacheable)?